### ISSUE TITLE
ttnn.gelu works incorrectly for inputs between -5.5 and -2.5

### ISSUE THREAD

**Issue by amalbasaTT** [tenstorrent/tt-metal#20562]
**Describe the bug**
ttnn.gelu works incorrectly for inputs between -5.5 and -2.5 and produces zeroes. Relative error for these inputs is between 94% and 100%, which is one of the causes of low accuracy in mixer_b16_224.goog_in21k model in pytorch2.0_ttnn project.

![Image](https://github.com/user-attachments/assets/a38a2d2f-6f05-4338-8ae5-2a8c0d53f192)

**To Reproduce**
Run unit test:
```
from loguru import logger
import random
import pytest
import torch
import ttnn
import traceback
from tests.ttnn.utils_for_testing import assert_with_pcc
aten = torch.ops.aten


def relative_error(expected, calculated):
    assert expected.shape == calculated.shape, "Tensors must have the same shape"
    if expected.dtype == torch.bool:
        expected = expected.to(torch.int)
    if calculated.dtype == torch.bool:
        calculated = calculated.to(torch.int)
    relative_error = torch.abs(expected - calculated) / (torch.abs(expected) + 1e-8)
    return relative_error


def run_gelu_tests(
    input_shape, 
    dtype, 
    dlayout,
    device,
):

    x = torch.Tensor(size=input_shape[0]).uniform_(-7.5, 0.0).to(torch.bfloat16)

    try:
        ref_value = aten.gelu.default(x, )
        
        tt_x = ttnn.from_torch(x, dtype=dtype[0], layout=dlayout[0], device=device)

        tt_result = ttnn.gelu(tt_x)
        tt_result = ttnn.to_torch(tt_result)
    except Exception as e:
        logger.warning(f"Test execution crashed: {e}")
        print(traceback.format_exc())
        raise e

    assert len(tt_result.shape) == len(ref_value.shape)
    assert tt_result.shape == ref_value.shape
    assert_with_pcc(ref_value, tt_result, 0.999)
    
    rel_error_tensor = relative_error(ref_value, tt_result)
    mre = torch.max(rel_error_tensor).item()
    inds = torch.where(rel_error_tensor > 0.3)
    
    assert torch.all(tt_result[inds] == 0), "Not all elements are zero"
    
    print(x[inds])

    assert mre < 0.3, f"Maximum relative error is: {mre}. Number of elements with that relative error is: {len(inds[0])}"

test_sweep_args = [
    (
        [(1, 768, 384)],
        [ttnn.bfloat16],
        [ttnn.TILE_LAYOUT],
    ),
    (
        [(1, 196, 3072)],
        [ttnn.bfloat16],
        [ttnn.TILE_LAYOUT],
    ),
]


@pytest.mark.parametrize(
    "input_shape, dtype, dlayout",
    (test_sweep_args),
)
def test_gelu(input_shape, dtype, dlayout, device):
    run_gelu_tests(input_shape, dtype, dlayout, device)

```

**Comment by cmaryanTT** [comment]
@mcw-vishal - can you please add some graphs of the old, new proposed and golden function; and a separate graph of the relative % error of the old and proposed implementations.

**Comment by mcw-vishal** [comment]
Sure @cmaryanTT

**Comment by cmaryanTT** [comment]
@mcw-vishal - from the first graph, it looks like we should switch to the linear y=0 approximation at around -4.5, which will give a more accurate approximation and will avoid the dip in the green line between -4.5 and -6. You could even search more accurately for the precise point where the approximations should switch over.

**Comment by mcw-vishal** [comment]
@amalbasaTT could you please check if the pcc now is as expected, from the changes in [branch](https://github.com/tenstorrent/tt-metal/tree/vishal/gelu_accurate_issue).

**Comment by cmaryanTT** [comment]
Thanks, this looks good. Just to confirm, the performance of the new version should be the same as the original? (similar computation, same polynomial order)

**Comment by mcw-vishal** [comment]
> Thanks, this looks good. Just to confirm, the performance of the new version should be the same as the original? (similar computation, same polynomial order)

Not necessarily since older version was only using one threshold, but in the new version we have three thresholds to check, I will note the performance data as well.

**Comment by mcw-vishal** [comment]
The profiling analysis suggests the newer approach is slower (approx twice) as compared to the older approach. It takes more time when the inputs are < 0 due to branching instructions for three thresholds. Please find below details:-

Parameter Analysed: DEVICE KERNEL DURATION (DKD)
Input values in the range [-6.0, 3.0]

1. Input Tensor Size: 10*1024 
    1. Older  DKD (in ns) :  10775
    2. Newer DKD (in ns) : 18965

1. Input Tensor Size: 1024*1024
    1. Older  DKD (in ns) :  866093
    2. Newer DKD (in ns) : 1722432

**Comment by cmaryanTT** [comment]
@mcw-vishal - we have to revisit the implementation. A 2x slowdown is a non-starter. Let's discuss this upcoming week.

**Comment by mcw-vishal** [comment]
The challenge is to fit the complex polynomial of gelu through a polynomial with single threshold.

- The RidgeRegression and Chebychev in polyval5, results are almost same.
- Rational approach(to avoid higher degree polynomial evaluation): computes numerator and denominator through polyval5 separately and tries to fit the polynomial as numerator/1+denominator. It involves two ploy5 evaluations and division.

Since we are fitting the polynomial in the range [-5.5, 3.0] and for values > 3.0 it behaves as identity, we can leverage this information to remove unnecessary sign checks for the positive cdf function. This will enable us to evaluate higher degree polynomials.

- Chebychev polyval15 gives the best result with cutoff range of [-5.5, 3.0]. But still the execution time is more. (~40% slower)

**Execution Time**
For tensor size 1024x1024

- Original gelu: 866093 ns
- ChebyChev Poly15: 1223033 ns (41% slower)
- Rational Poly5: 1465826 ns (69% slower)

Attaching ULP graphs in the following order.

1. ChebyChev Poly15 ULP graph
2. Rational Approach ULP graph
3. Default ttnn.gelu ULP graph

![Image](https://github.com/user-attachments/assets/ad5cc412-bb57-4fc6-8368-1f4e8ee586c1)

![Image](https://github.com/user-attachments/assets/d9d374ba-7bd7-4842-8bd2-b2868a4c4b3a)

![Image](https://github.com/user-attachments/assets/5868743d-96bf-460d-985e-a79feee6c771)

**Comment by mcw-vishal** [comment]
Adding abs diff and accuracy plots. 

![Image](https://github.com/user-attachments/assets/13d37e35-a769-42ba-8e57-9db640e4b87f)
![Image](https://github.com/user-attachments/assets/3b69bf38-c20a-4e1d-87b3-1dced2f9ba3e)

**Comment by mcw-vishal** [comment]
Issue will be solved by [PR](https://github.com/tenstorrent/tt-metal/pull/27283).

**Comment by mcw-vishal** [comment]
Closing as  [#27283](https://github.com/tenstorrent/tt-metal/pull/27283) is merged in main.
